infer/inference.cc

```diff
@@ -269,6 +269,14 @@ unique_ptr<cfg::CFG> Inference::run(core::Context ctx, unique_ptr<cfg::CFG> cfg)
         for (auto &pair : current.vars()) {
             pair.second.knowledge.emitKnowledgeSizeMetric();
         }
+
+        core::TypePtr conditionalType = current.getTypeAndOrigin(ctx, bb->bexit.cond.variable).type;
+
+        if (!conditionalType.isUntyped() && core::Types::isSubType(ctx, conditionalType, core::Types::void_())) {
```

Comment:
i think you've got this backwards (you want void < conditionalType)

This will only catch the case where the branch variable's type is void (or some subtype of void, which there are none)

you want to raise an error if any component of the type is void. here's your test case:

```ruby
extend T::Sig

sig {void}
def returns_void; end

if T.unsafe(nil)
  x = returns_void
else
  x = false
end

if x
  T.reveal_type(x)
else
  T.reveal_type(x)
end
```

Reply:
Indeed it was, although this particular example worked with the backwards version too.

Reply:
Wait really? In that case, can you look for a test that would have failed had the order not been swapped?

Reply:
> Wait really? In that case, can you look for a test that would have failed had the order not been swapped?

Here is a better test case:

```ruby
# typed: true

class D
  extend T::Sig
  sig {void}
  def returns_void; end

  def test
    if T.unsafe(nil)
      x = returns_void
    else
      x = false
    end

    puts 'defeat simiplification'

    if x
      T.reveal_type(x)
    else
      T.reveal_type(x)
    end
  end
end
```

as i wrote it before it happened to work because Sorbet simplified the CFG into an equivalent program that had slightly different behavior

Reply:
(this test does not report an error but should have had the order of `conditionalType` and `void` been the other way around)